### PR TITLE
`azurerm_windows_function_app`, `azurerm_linux_function_app` - fix `WEBSITE_RUN_FROM_PACKAGE` being removed from `app_setting` issue

### PR DESCRIPTION
**Pull request by xiaxyi**
This is a hot discussed issue. 

When function app is deployed via other platform such as azure devOps, or when function app is a zip deployment. The `WEBSITE_RUN_FROM_PACKAGE` will be added outside of Terraform, so the kvp will be removed when user doing terraform apply and the removal will cause function app to be broke.

fix https://github.com/hashicorp/terraform-provider-azurerm/issues/17081

For linux function app, the `WEBSITE_RUN_FROM_PACKAGE` is not be written back to state if terraform provider doesn't find it in config so ignore change is not going to work as no diff will be detected between state and config, thus `WEBSITE_RUN_FROM_PACKAGE` will still be removed.

For windows function app, we set back the `WEBSITE_RUN_FROM_PACKAGE` and don't do the config check by `metadata.get(app_setting.WEBSITE_RUN_FROM_PACKAGE)` so ignore change works. 

In this PR, a new property `website_run_from_package` is added, the default value is set to false, so for current user, they will get below diff when running the plan but they won't lose the `WEBSITE_RUN_FROM_PACKAGE` in `app_setting` as we set the value back
```
~ resource "azurerm_windows_function_app" "test" {
        id                                = "/subscriptions/xxx/resourceGroups/xiaxin-func-deploy-rg/providers/Microsoft.Web/sites/xiaxin-func-deploy-windows"
        name                              = "xiaxin-func-deploy-windows"
        tags                              = {}
      ~ website_run_from_package          = true -> false
        # (23 unchanged attributes hidden)

        # (1 unchanged block hidden)
    }
```

In this solution, user doesn't need to use the ignore change for `WEBSITE_RUN_FROM_PACKAGE` as they can explicitly set the website_run_from_package to true. When terraform reads the instance, it will check app_setting to see if `WEBSITE_RUN_FROM_PACKAGE` is there then set the property `website_run_from_package` to `true`.

### CODE REVIEW

**Comment by xiaxyi** [comment]
Example usage:
```
resource "azurerm_linux_function_app" "test" {
    name = "functuion_name"
    resource_group_name = "rg_name"
    location = "east us"
    service_plan_id = "service_plan_id"
    storage_account_name = "saaccount"
    
    website_run_from_package = true
    site_config{
        application_stack {
          dotnet_version = "6.0"
        }
    }

    app_settings = {
      "terraformkey" = "terraformvalue"
    }
}
```

File: website/docs/d/linux_function_app.html.markdown [hunk 1]
```diff
@@ -101,6 +101,8 @@ In addition to the Arguments listed above - the following Attributes are exporte
 
 * `storage_uses_managed_identity` - Does the Function App use Managed Identity to access the storage account?
 
+* `website_run_from_package` - Does the Function App deployed via a package file?
```
Comment (jackofallops) [inline website/docs/d/linux_function_app.html.markdown:104]:
```suggestion
* `website_run_from_package` - Is the Linux Function App deployed via a package file? 
```

File: website/docs/d/windows_function_app.html.markdown [hunk 2]
```diff
@@ -101,6 +101,8 @@ In addition to the Arguments listed above - the following Attributes are exporte
 
 * `storage_uses_managed_identity` - Is the Function App using a Managed Identity to access the storage account?
 
+* `website_run_from_package` - Does the Function App deployed via a package file?
```
Comment (jackofallops) [inline website/docs/d/windows_function_app.html.markdown:104]:
```suggestion
* `website_run_from_package` - Is the Windows Function App deployed via a package file?
```

File: website/docs/r/linux_function_app.html.markdown [hunk 3]
```diff
@@ -123,6 +123,8 @@ The following arguments are supported:
 
 ~> **NOTE:** `storage_key_vault_secret_id` used without a version will use the latest version of the secret, however, the service can take up to 24h to pick up a rotation of the latest version. See the [official docs](https://docs.microsoft.com/azure/app-service/app-service-key-vault-references#rotation) for more information.
 
+* `website_run_from_package` - (Optional) Does the function App deployed from a package file? Defaults to `false`.
```
Comment (jackofallops) [inline website/docs/r/linux_function_app.html.markdown:126]:
```suggestion
* `website_run_from_package` - (Optional) Should the Linux Function App be deployed from a package file? Defaults to `false`.
```

File: website/docs/r/linux_function_app.html.markdown [hunk 4]
```diff
@@ -133,7 +135,7 @@ The following arguments are supported:
 
 * `zip_deploy_file` - (Optional) The local path and filename of the Zip packaged application to deploy to this Linux Function App.
 			
-~> **Note:** Using this value requires either `WEBSITE_RUN_FROM_PACKAGE=1` or `SCM_DO_BUILD_DURING_DEPLOYMENT=true` to be set on the App in `app_settings`. Refer to the [Azure docs](https://learn.microsoft.com/en-us/azure/azure-functions/functions-deployment-technologies) for further details.
+~> **Note:** Using this value requires either `WEBSITE_RUN_FROM_PACKAGE=1` or `SCM_DO_BUILD_DURING_DEPLOYMENT=true` to be set on the App in `app_settings`. For Linux function app running in consumption plan, please set a URL that is the remote location of the specific package file you want to run for `WEBSITE_RUN_FROM_PACKAGE` in `app_setting`. Refer to the [Azure docs](https://learn.microsoft.com/en-us/azure/azure-functions/functions-deployment-technologies) for further details.
```
Comment (jackofallops) [inline website/docs/r/linux_function_app.html.markdown:138]:
```suggestion
~> **Note:** Using this value requires either `WEBSITE_RUN_FROM_PACKAGE=1` or `SCM_DO_BUILD_DURING_DEPLOYMENT=true` to be set on the App in `app_settings`, or the `website_run_from_package` property to be set to `true`. For Linux Function Apps running in a Consumption Service Plan, a URL pointing to the remote location of the package file to be run must be specified for  `WEBSITE_RUN_FROM_PACKAGE` in `app_setting`. Refer to the [Azure docs](https://learn.microsoft.com/en-us/azure/azure-functions/functions-deployment-technologies) for further details.
```

File: website/docs/r/windows_function_app.html.markdown [hunk 5]
```diff
@@ -135,6 +135,8 @@ The following arguments are supported:
 			
 ~> **Note:** Using this value requires `WEBSITE_RUN_FROM_PACKAGE=1` to be set on the App in `app_settings`. Refer to the [Azure docs](https://learn.microsoft.com/en-us/azure/azure-functions/functions-deployment-technologies) for further details.
 
+* `website_run_from_package` - (Optional) Does the function App deployed from a package file? Defaults to `false`.
```
Comment (jackofallops) [inline website/docs/r/windows_function_app.html.markdown:138]:
```suggestion
* `website_run_from_package` - (Optional) Should the Windows Function App be deployed from a package file? Defaults to `false`.
```

File: internal/services/appservice/windows_function_app_resource.go [hunk 6]
```diff
@@ -1263,7 +1288,13 @@ func (m *WindowsFunctionAppModel) unpackWindowsFunctionAppSettings(input web.Str
 			m.BuiltinLogging = true
 
 		case "WEBSITE_VNET_ROUTE_ALL":
-			// Filter out - handled by site_config setting `vnet_route_all_enabled`
+		// Filter out - handled by site_config setting `vnet_route_all_enabled`
+
+		case "WEBSITE_RUN_FROM_PACKAGE":
+			// Keep if user explicitly set, otherwise filter out as will have been added by ADO et al
+			if _, ok := metadata.ResourceData.GetOk("app_settings.WEBSITE_RUN_FROM_PACKAGE"); ok {
+				appSettings[k] = utils.NormalizeNilableString(v)
```
Comment (jackofallops) [inline internal/services/appservice/windows_function_app_resource.go:1296]: The utils package is going to be gradually deprecated, so `pointer.From()` should be used here.
```suggestion
				appSettings[k] = pointer.From(v)
```

File: internal/services/appservice/windows_function_app_resource.go [hunk 7]
```diff
@@ -716,6 +730,9 @@ func (r WindowsFunctionAppResource) Read() sdk.ResourceFunc {
 
 			state.unpackWindowsFunctionAppSettings(appSettingsResp, metadata)
 
+			if appSettingsResp.Properties != nil && appSettingsResp.Properties["WEBSITE_RUN_FROM_PACKAGE"] != nil {
+				state.WebsiteRunFromPackage = true
+			}
```
Comment (jackofallops) [inline internal/services/appservice/windows_function_app_resource.go:735]: as above
```suggestion
			if appSettingsResp.Properties != nil && pointer.From(appSettingsResp.Properties["WEBSITE_RUN_FROM_PACKAGE"]) != "" {
				state.WebsiteRunFromPackage = true
			}
```

File: internal/services/appservice/windows_function_app_resource.go [hunk 8]
```diff
@@ -945,6 +962,14 @@ func (r WindowsFunctionAppResource) Update() sdk.ResourceFunc {
 				}
 			}
 
+			if appSettingsResp.Properties != nil && appSettingsResp.Properties["WEBSITE_RUN_FROM_PACKAGE"] != nil {
+				if state.AppSettings == nil {
+					state.AppSettings = make(map[string]string)
+				}
+				state.WebsiteRunFromPackage = true
+				state.AppSettings["WEBSITE_RUN_FROM_PACKAGE"] = *appSettingsResp.Properties["WEBSITE_RUN_FROM_PACKAGE"]
+			}
```
Comment (jackofallops) [inline internal/services/appservice/windows_function_app_resource.go:971]: as above
```suggestion
			if appSettingsResp.Properties != nil && pointer.From(appSettingsResp.Properties["WEBSITE_RUN_FROM_PACKAGE"]) != "" {
				if state.AppSettings == nil {
					state.AppSettings = make(map[string]string)
				}
				state.WebsiteRunFromPackage = true
				state.AppSettings["WEBSITE_RUN_FROM_PACKAGE"] = *appSettingsResp.Properties["WEBSITE_RUN_FROM_PACKAGE"]
			}
```

**Review by jackofallops — CHANGES_REQUESTED** [review]
Hi @xiaxyi - Thanks for this, it's looking mostly good, just a few comments below. If you can take a look, we can continue review. Are you able to do any manual testing for this since we're relying on an external change to know this works?

Thanks

**Comment by xiaxyi** [comment]
Thanks @jackofallops for the code review. I updated the code per your suggestion.

I did the tests about external deployment to function app and the `WEBSITE_RUN_FROM_PACKAGE` will be added during the deployment. Is this the manual tests you were referring to?

**Comment by dotSource-bke** [comment]
Hello @jackofallops, @xiaxyi,
is there a reason why the pull request is stuck for such a long time at that stage of development? If so, please provide an update so it is possible to plan when a provider that is fixed will be available. Otherwise, please just fix this painful issue.

**Review by jackofallops — CHANGES_REQUESTED** [review]
Hi @xiaxyi - In re-review, I think the problem here is actually a bug here: https://github.com/hashicorp/terraform-provider-azurerm/blob/main/internal/services/appservice/linux_function_app_resource.go#L1291-L1295
This will be preventing the `ignore_change` from working I believe. Removing this handling  will allow users to correctly `ignore_change` which would be the desirable behaviour here, rather than introducing further configuration / complication.  

Specifically, if these lines are removed, the following will correctly step over the external changes to the `app_settings` map:

```terraform
lifecycle {
    ignore_changes = [ app_settings["WEBSITE_RUN_FROM_PACKAGE"] ]
  }
```
(note: this is the correct way to reference the map item, not the dotted notation - it may be worth also including this in the docs?)

This already works for the `azurerm_windows_function_app` resource (I just tested it by hand)

If you can take a look and update the PR accordingly, I think this should be able to go in.

Thanks!

**Comment by katbyte** [comment]
@xiaxyi as this has been untouched for ~3 months i am going to close it out. Please do feel free to reopen it once you are ready to continue working on it. thanks!

**Pull request closed without merging**